File: scripts/convert_public_input.py
import json
import base58
import argparse

def is_uint256(value):
    """Check if the string is a uint256 value in hex."""
    return isinstance(value, str) and value.startswith("0x") and len(value) == 66

def is_hash(value):
    """Check assuming hash values don't start with '0x' and are strings longer than 40 characters."""
    return isinstance(value, str) and not value.startswith("0x") and len(value) > 40

def is_integer_string(value):
    """Check if the string represents an integer."""
    try:
        int(value)
        return True
    except:
        return False
    
def convert_to_uint256(node):
    """Recursively parse a JSON node and convert to uint256."""
    uint256_values = []
    
    if isinstance(node, dict):
        for key, value in node.items():
            uint256_values.extend(convert_to_uint256(value))
    elif isinstance(node, list):
        for item in node:
            uint256_values.extend(convert_to_uint256(item))
    else:
        # Leaf node
        if is_uint256(node):
            uint256_values.append(int(node, 16))
        elif is_hash(node):
            # TODO: 255 or 256?
            uint256_values.append(base58.b58decode_int(node) % (2**256))
        elif is_integer_string(node):
            num = int(node)
            # Encoding the sign and value into two uint256 numbers
            if num < 0:
                uint256_values.append(1)
                uint256_values.append(abs(num))
            else:
                uint256_values.append(0)
                uint256_values.append(num)
        elif isinstance(node, bool):
            uint256_values.append(1 if node else 0)
    
    return uint256_values

def convert_from_uint256(node, uint256_values):
    """Recursively parse an exemplary JSON node and convert from uint256."""
    if isinstance(node, dict):
        new_node = {}
        for key, value in node.items():
            new_node[key], uint256_values = convert_from_uint256(value, uint256_values)
        return new_node, uint256_values
    elif isinstance(node, list):
        new_node = []
        for item in node:
            converted_item, uint256_values = convert_from_uint256(item, uint256_values)
            new_node.append(converted_item)
        return new_node, uint256_values
    else:
        # Leaf node
        if is_uint256(node):
            return "0x{:064x}".format(uint256_values.pop(0)), uint256_values
        elif is_hash(node):
            num = uint256_values.pop(0)
            return base58.b58encode(num.to_bytes(32, byteorder='big')).decode('utf-8'), uint256_values
        elif is_integer_string(node):
            sign_bit = uint256_values.pop(0)
            num = uint256_values.pop(0)
            return str(-num if sign_bit else num), uint256_values
        elif isinstance(node, bool):
            return bool(uint256_values.pop(0)), uint256_values
        else:
            return node, uint256_values
        
def convert_to_array_parser(args):
    with open(args.input) as f:
        input = json.load(f)
    uint256_values = convert_to_uint256(input)
    with open(args.output, "w") as f:
        json.dump(uint256_values, f, indent=4)

def convert_to_json_parser(args):
    with open(args.input) as f:
        uint256_values = json.load(f)
    with open(args.example) as f:
        example = json.load(f)
    output, _ = convert_from_uint256(example, uint256_values)
    with open(args.output, "w") as f:
        json.dump(output, f, indent=4)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        prog='Helper for =nil; Proof Market Endpoint',
        description='Changes format of public input for =nil; Proof Market endpoint')
    parser.add_argument('--input', help="Input file path", default="input.json")
    parser.add_argument('--output', help="Output file path", default="output.json")

    subparsers = parser.add_subparsers(help="sub-command help")
    parser_to_JSON = subparsers.add_parser("from_uint",
                                           help="Conver public input from uint256[] to JSON format")
    parser_to_JSON.add_argument('--example', help="Example public input file path", default="data/example.json")
    parser_to_JSON.set_defaults(func=convert_to_json_parser)

    parser_to_array = subparsers.add_parser("from_json", help="Convert public input from JSON to uint256[]")
    parser_to_array.set_defaults(func=convert_to_array_parser)

    args = parser.parse_args()
    args.func(args)

File: scripts/convert_public_input.js
const fs = require('fs');
const path = require('path');
const process = require('process');
const base58 = require('base-x')('123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz');
const yargs = require('yargs');
const { hideBin } = require('yargs/helpers');
const LosslessJSON = require('lossless-json');

function isUint256(value) {
    return typeof value === "string" && value.startsWith("0x") && value.length === 66;
}

function isHash(value) {
    return typeof value === "string" && !value.startsWith("0x") && value.length > 40;
}

function isIntegerString(value) {
    // value is a string that represents an integer, e.g. "123" or "-123"
    return typeof value === "string" && !value.startsWith("0x") && !isNaN(Number(value));
}

function convertToUint256(node) {
    let uint256Values = [];

    if (node instanceof Object && !(node instanceof Array)) {
        for (const key in node) {
            uint256Values = uint256Values.concat(convertToUint256(node[key]));
        }
    } else if (node instanceof Array) {
        node.forEach(item => {
            uint256Values = uint256Values.concat(convertToUint256(item));
        });
    } else {
        if (isUint256(node)) {
            uint256Values.push(BigInt(node));
            // uint256Values.push('0x' + BigInt(node).toString(16).padStart(64, '0'));
        } else if (isHash(node)) {
            const num = base58.decode(node);
            uint256Values.push(Number(num) % (2**256));
        } else if (isIntegerString(node)) {
            const num = Number(node);
            if (num < 0) {
                uint256Values.push(1, Math.abs(num));
            } else {
                uint256Values.push(0, num);
            }
        } else if (typeof node === "boolean") {
            uint256Values.push(node ? 1 : 0);
        }
    }
    return uint256Values;
}

function convertFromUint256(node, uint256Values) {
    if (node instanceof Object && !(node instanceof Array)) {
        const newNode = {};
        for (const key in node) {
            newNode[key] = convertFromUint256(node[key], uint256Values);
        }
        return newNode;
    } else if (node instanceof Array) {
        return node.map(item => convertFromUint256(item, uint256Values));
    } else {
        if (isUint256(node)) {
            let uint256Value = BigInt(uint256Values.shift());
            let converted = '0x' + uint256Value.toString(16).padStart(64, '0');
            return converted;
        } else if (isHash(node)) {
            const num = BigInt(uint256Values.shift());
            const buffer = Buffer.alloc(32); // 32 bytes for a uint256
            for (let i = 0; i < 32; i++) {
                buffer[31 - i] = Number((num >> BigInt(i * 8)) & BigInt(0xFF));
            }
            return base58.encode(buffer);
        } else if (isIntegerString(node)) {
            const signBit = uint256Values.shift();
            const num = uint256Values.shift();
            return signBit === 1 ? String(-num) : String(num);
        } else if (typeof node === "boolean") {
            return Boolean(uint256Values.shift());
        } else {
            return node;
        }
    }
}

async function convertToArray(inputPath, outputPath) {
    const input = JSON.parse(fs.readFileSync(inputPath, 'utf-8'));
    const uint256Values = convertToUint256(input);
    console.log('uint256Values', uint256Values)
    fs.writeFileSync(outputPath, LosslessJSON.stringify(uint256Values, null, 4));
}

async function convertToJson(inputPath, outputPath, examplePath) {
    const uint256Values = LosslessJSON.parse(fs.readFileSync(inputPath, 'utf-8'));
    const exampleJson = JSON.parse(fs.readFileSync(examplePath, 'utf-8'));
    const convertedJson = convertFromUint256(exampleJson, uint256Values);
    fs.writeFileSync(outputPath, JSON.stringify(convertedJson, null, 4));
}

if (require.main === module) {
    const argv = yargs(hideBin(process.argv))
    .option('input', {
        type: 'string',
        description: 'Input file path',
        default: 'input.json',
    })
    .option('output', {
        type: 'string',
        description: 'Output file path',
        default: 'output.json',
    })
    .command(
        'from_json',
        'Convert public input from JSON to uint256[]',
        {
            input: {
                type: 'string',
                demandOption: true,
                description: 'Input file path for JSON public input',
            },
            output: {
                type: 'string',
                demandOption: true,
                description: 'Output file path for uint256 array',
            },
        },
        (argv) => {
            convertToArray(argv.input, argv.output)
                .then(() => process.exit(0))
                .catch((error) => {
                    console.error(error);
                    process.exit(1);
                });
        }
    )
    .command(
        'from_uint',
        'Convert public input from uint256[] to JSON format',
        {
            input: {
                type: 'string',
                demandOption: true,
                description: 'Input file path for uint256 array',
            },
            output: {
                type: 'string',
                demandOption: true,
                description: 'Output file path for JSON public input',
            },
            example: {
                type: 'string',
                demandOption: true,
                description: 'Example public input file path',
            },
        },
        (argv) => {
            convertToJson(argv.input, argv.output, argv.example)
                .then(() => process.exit(0))
                .catch((error) => {
                    console.error(error);
                    process.exit(1);
                });
        }
    )
    .demandCommand(1, 'You need to specify a command')
    .help()
    .argv;
}


module.exports = {
    convertToUint256,
    convertFromUint256,
    convertToArray,
    convertToJson,
};